<html>
<head>
  <title>VEP summary</title>
  <script type="text/javascript" src="http://www.google.com/jsapi"></script>
  <script type="text/javascript">
    google.load('visualization', '1', {packages: ['corechart','table']});
  </script>
  <script type="text/javascript">
    
    function init() {
      // charts
      var var_class_pie = drawPie('var_class_pie', 'Variant classes', google.visualization.arrayToDataTable([['Variant class','Count'],['indel',767],['sequence_alteration',1082],['insertion',7418],['deletion',7855],['SNV',80796]]), null);
var var_class_table = drawTable('var_class_table', 'Variant classes', google.visualization.arrayToDataTable([['Variant class','Count'],['indel',767],['sequence_alteration',1082],['insertion',7418],['deletion',7855],['SNV',80796]]));

            google.visualization.events.addListener(var_class_pie, 'select', function() {
              var_class_table.setSelection(var_class_pie.getSelection());
            });
            google.visualization.events.addListener(var_class_table, 'select', function() {
              var_class_pie.setSelection(var_class_table.getSelection());
            });
          var var_cons_pie = drawPie('var_cons_pie', 'Consequences (most severe)', google.visualization.arrayToDataTable([['Consequence type','Count'],['splice_donor_variant',13],['splice_acceptor_variant',17],['stop_gained',11],['frameshift_variant',24],['stop_lost',1],['start_lost',2],['inframe_insertion',11],['inframe_deletion',14],['missense_variant',554],['protein_altering_variant',1],['splice_region_variant',167],['incomplete_terminal_codon_variant',1],['synonymous_variant',472],['mature_miRNA_variant',1],['5_prime_UTR_variant',410],['3_prime_UTR_variant',1653],['non_coding_transcript_exon_variant',2927],['intron_variant',50626],['upstream_gene_variant',10128],['downstream_gene_variant',6211],['TF_binding_site_variant',63],['regulatory_region_variant',1911],['intergenic_variant',22700]]), {slices: [{color: "#ff7f50"}, {color: "#ff7f50"}, {color: "#ff0000"}, {color: "#ff69b4"}, {color: "#ff0000"}, {}, {color: "#ff69b4"}, {color: "#ff69b4"}, {color: "#ffd700"}, {}, {color: "#ff7f50"}, {color: "#ff00ff"}, {color: "#76ee00"}, {}, {}, {}, {}, {color: "#02599c"}, {color: "#a2b5cd"}, {color: "#a2b5cd"}, {}, {color: "brown"}, {color: "gray"}]});
var var_cons_table = drawTable('var_cons_table', 'Consequences (most severe)', google.visualization.arrayToDataTable([['Consequence type','Count'],['splice_donor_variant',13],['splice_acceptor_variant',17],['stop_gained',11],['frameshift_variant',24],['stop_lost',1],['start_lost',2],['inframe_insertion',11],['inframe_deletion',14],['missense_variant',554],['protein_altering_variant',1],['splice_region_variant',167],['incomplete_terminal_codon_variant',1],['synonymous_variant',472],['mature_miRNA_variant',1],['5_prime_UTR_variant',410],['3_prime_UTR_variant',1653],['non_coding_transcript_exon_variant',2927],['intron_variant',50626],['upstream_gene_variant',10128],['downstream_gene_variant',6211],['TF_binding_site_variant',63],['regulatory_region_variant',1911],['intergenic_variant',22700]]));

            google.visualization.events.addListener(var_cons_pie, 'select', function() {
              var_cons_table.setSelection(var_cons_pie.getSelection());
            });
            google.visualization.events.addListener(var_cons_table, 'select', function() {
              var_cons_pie.setSelection(var_cons_table.getSelection());
            });
          var consequences_pie = drawPie('consequences_pie', 'Consequences (all)', google.visualization.arrayToDataTable([['Consequence type','Count'],['splice_donor_variant',28],['splice_acceptor_variant',32],['stop_gained',17],['frameshift_variant',51],['stop_lost',1],['start_lost',3],['inframe_insertion',26],['inframe_deletion',21],['protein_altering_variant',1],['missense_variant',1377],['splice_region_variant',534],['incomplete_terminal_codon_variant',1],['synonymous_variant',1387],['coding_sequence_variant',2],['mature_miRNA_variant',1],['5_prime_UTR_variant',1000],['3_prime_UTR_variant',3194],['non_coding_transcript_exon_variant',5323],['intron_variant',248514],['NMD_transcript_variant',22208],['non_coding_transcript_variant',67006],['upstream_gene_variant',63915],['downstream_gene_variant',60244],['TF_binding_site_variant',606],['regulatory_region_variant',11952],['intergenic_variant',25456]]), {slices: [{color: "#ff7f50"}, {color: "#ff7f50"}, {color: "#ff0000"}, {color: "#ff69b4"}, {color: "#ff0000"}, {}, {color: "#ff69b4"}, {color: "#ff69b4"}, {}, {color: "#ffd700"}, {color: "#ff7f50"}, {color: "#ff00ff"}, {color: "#76ee00"}, {color: "#458b00"}, {}, {}, {}, {}, {color: "#02599c"}, {}, {}, {color: "#a2b5cd"}, {color: "#a2b5cd"}, {}, {color: "brown"}, {color: "gray"}]});
var consequences_table = drawTable('consequences_table', 'Consequences (all)', google.visualization.arrayToDataTable([['Consequence type','Count'],['splice_donor_variant',28],['splice_acceptor_variant',32],['stop_gained',17],['frameshift_variant',51],['stop_lost',1],['start_lost',3],['inframe_insertion',26],['inframe_deletion',21],['protein_altering_variant',1],['missense_variant',1377],['splice_region_variant',534],['incomplete_terminal_codon_variant',1],['synonymous_variant',1387],['coding_sequence_variant',2],['mature_miRNA_variant',1],['5_prime_UTR_variant',1000],['3_prime_UTR_variant',3194],['non_coding_transcript_exon_variant',5323],['intron_variant',248514],['NMD_transcript_variant',22208],['non_coding_transcript_variant',67006],['upstream_gene_variant',63915],['downstream_gene_variant',60244],['TF_binding_site_variant',606],['regulatory_region_variant',11952],['intergenic_variant',25456]]));

            google.visualization.events.addListener(consequences_pie, 'select', function() {
              consequences_table.setSelection(consequences_pie.getSelection());
            });
            google.visualization.events.addListener(consequences_table, 'select', function() {
              consequences_pie.setSelection(consequences_table.getSelection());
            });
          var coding_pie = drawPie('coding_pie', 'Coding consequences', google.visualization.arrayToDataTable([['Consequence type','Count'],['stop_gained',17],['frameshift_variant',51],['stop_lost',1],['inframe_insertion',26],['inframe_deletion',21],['missense_variant',1377],['synonymous_variant',1387],['coding_sequence_variant',2]]), {slices: [{color: "#ff0000"}, {color: "#ff69b4"}, {color: "#ff0000"}, {color: "#ff69b4"}, {color: "#ff69b4"}, {color: "#ffd700"}, {color: "#76ee00"}, {color: "#458b00"}]});
var coding_table = drawTable('coding_table', 'Coding consequences', google.visualization.arrayToDataTable([['Consequence type','Count'],['stop_gained',17],['frameshift_variant',51],['stop_lost',1],['inframe_insertion',26],['inframe_deletion',21],['missense_variant',1377],['synonymous_variant',1387],['coding_sequence_variant',2]]));

            google.visualization.events.addListener(coding_pie, 'select', function() {
              coding_table.setSelection(coding_pie.getSelection());
            });
            google.visualization.events.addListener(coding_table, 'select', function() {
              coding_pie.setSelection(coding_table.getSelection());
            });
          var sift_pie = drawPie('sift_pie', 'SIFT summary', google.visualization.arrayToDataTable([['Prediction','Count'],['deleterious - low confidence',58],['tolerated - low confidence',156],['deleterious',268],['tolerated',798]]), {slices: [{}, {}, {color: "red"}, {color: "green"}]});
var sift_table = drawTable('sift_table', 'SIFT summary', google.visualization.arrayToDataTable([['Prediction','Count'],['deleterious - low confidence',58],['tolerated - low confidence',156],['deleterious',268],['tolerated',798]]));

            google.visualization.events.addListener(sift_pie, 'select', function() {
              sift_table.setSelection(sift_pie.getSelection());
            });
            google.visualization.events.addListener(sift_table, 'select', function() {
              sift_pie.setSelection(sift_table.getSelection());
            });
          var polyphen_pie = drawPie('polyphen_pie', 'PolyPhen summary', google.visualization.arrayToDataTable([['Prediction','Count'],['unknown',76],['possibly damaging',131],['probably damaging',133],['benign',1037]]), {slices: [{color: "blue"}, {color: "orange"}, {color: "red"}, {color: "green"}]});
var polyphen_table = drawTable('polyphen_table', 'PolyPhen summary', google.visualization.arrayToDataTable([['Prediction','Count'],['unknown',76],['possibly damaging',131],['probably damaging',133],['benign',1037]]));

            google.visualization.events.addListener(polyphen_pie, 'select', function() {
              polyphen_table.setSelection(polyphen_pie.getSelection());
            });
            google.visualization.events.addListener(polyphen_table, 'select', function() {
              polyphen_pie.setSelection(polyphen_table.getSelection());
            });
          var chr_bar = drawBar('chr_bar', 'Variants by chromosome', google.visualization.arrayToDataTable([['Chromosome','Count'],['22',97918]]), {legend: {position: "none"}});
var chr_table = drawTable('chr_table', 'Variants by chromosome', google.visualization.arrayToDataTable([['Chromosome','Count'],['22',97918]]));

            google.visualization.events.addListener(chr_bar, 'select', function() {
              chr_table.setSelection(chr_bar.getSelection());
            });
            google.visualization.events.addListener(chr_table, 'select', function() {
              chr_bar.setSelection(chr_table.getSelection());
            });
          var chr_22_area = drawArea('chr_22_area', 'Distribution of variants on chromosome 22', google.visualization.arrayToDataTable([['Position (mb)','Count'],['0',0],['1',0],['2',0],['3',0],['4',0],['5',0],['6',0],['7',0],['8',0],['9',0],['10',0],['11',0],['12',0],['13',0],['14',0],['15',0],['16',5207],['17',4434],['18',4024],['19',2800],['20',2506],['21',3326],['22',3380],['23',3605],['24',1972],['25',3268],['26',2260],['27',2906],['28',1150],['29',1942],['30',1775],['31',2354],['32',2444],['33',2374],['34',2791],['35',2721],['36',2071],['37',2742],['38',1943],['39',2641],['40',1379],['41',2014],['42',2555],['43',3162],['44',2981],['45',3514],['46',1801],['47',2778],['48',2679],['49',4637],['50',3042],['51',740]]), {hAxis: {title: "Position (mb)", textStyle: {fontSize: 8}}, legend: {position: "none"}});
var protein_bar = drawBar('protein_bar', 'Position in protein', google.visualization.arrayToDataTable([['Position in protein (percentile)','Count'],['00-10%',274],['10-20%',280],['20-30%',292],['30-40%',248],['40-50%',309],['50-60%',254],['60-70%',288],['70-80%',287],['80-90%',336],['90-100%',316]]), {hAxis: {title: "Position in protein (percentile)", textStyle: {fontSize: 10}}, legend: {position: "none"}});

    }
    
    function drawPie(id, title, data, options) {    
      var pie = new google.visualization.PieChart(document.getElementById(id));
      pie.draw(data, options);
      return pie;
    }
    function drawBar(id, title, data, options) {
      var bar = new google.visualization.ColumnChart(document.getElementById(id));
      bar.draw(data, options);
      return bar;
    }
    function drawTable(id, title, data) {
      var table = new google.visualization.Table(document.getElementById(id));
      table.draw(data, null);
      return table;
    }
    function drawLine(id, title, data, options) {
      var line = new google.visualization.LineChart(document.getElementById(id));
      line.draw(data, options);
      return line;
    }
    function drawArea(id, title, data, options) {
      var area = new google.visualization.AreaChart(document.getElementById(id));
      area.draw(data, options);
      return area;
    }
    google.setOnLoadCallback(init);
  </script>
  
  
  <style type="text/css">
    body {
      font-family: arial, sans-serif;
      margin: 0px;
      padding: 0px;
    }
    
    a {color: #36b;}
    a.visited {color: #006;}
    
    .stats_table {
      margin: 5px;
    }
    
    tr:nth-child(odd) {
      background-color: rgb(238, 238, 238);
    }
    
    td {
      padding: 5px;
    }
    
    td:nth-child(odd) {
      font-weight: bold;
    }
    
    h3 {
      color: #666;
    }
    
    .masthead {
      background-color: rgb(51, 51, 102);
      color: rgb(204, 221, 255);
      height: 80px;
      width: 100%;
      padding: 0px;
    }
    
    .main {
      padding: 10px;
    }
    
    .gradient {
      background: #333366; /* Old browsers */
      background: -moz-linear-gradient(left,  #333366 0%, #ffffff 100%); /* FF3.6+ */
      background: -webkit-gradient(linear, left top, right top, color-stop(0%,#333366), color-stop(100%,#ffffff)); /* Chrome,Safari4+ */
      background: -webkit-linear-gradient(left,  #333366 0%,#ffffff 100%); /* Chrome10+,Safari5.1+ */
      background: -o-linear-gradient(left,  #333366 0%,#ffffff 100%); /* Opera 11.10+ */
      background: -ms-linear-gradient(left,  #333366 0%,#ffffff 100%); /* IE10+ */
      background: linear-gradient(to right,  #333366 0%,#ffffff 100%); /* W3C */
      filter: progid:DXImageTransform.Microsoft.gradient( startColorstr='#333366', endColorstr='#ffffff',GradientType=1 ); /* IE6-9 */
      
      padding: 0px;
      height: 80px;
      width: 500px;
    }
    
    .main_content {
      margin-left: 300px;
    }
    
    .sidemenu {
      width: 260px;
      position: fixed;
      border-style: solid;
      border-width: 2px;
      border-color: rgb(51, 51, 102);
    }
    
    .sidemenu_head {
      width: 250px;
      background-color: rgb(51, 51, 102);
      color: rgb(204, 221, 255);
      padding: 5px;
    }
    
    .sidemenu_body {
      width: 250px;
      padding: 5px;
    }
  </style>
</head>
<body>
<div id="masthead" class="masthead">
  <div style="float: left; display: inline; padding: 10px; height: 80px;">
    <a href="http://www.ensembl.org/"><img src="http://static.ensembl.org/i/e-ensembl.png"></a>
  </div>
  
  <div style="float: right; display: inline; height: 80px; background: white; padding: 10px;">
    <a href="http://www.ensembl.org/info/docs/variation/vep/vep_script.html"><img src="http://www.ensembl.org/img/vep_logo.png"></a>
  </div>
  <div class="gradient">
  </div>
</div>
<div class="main">
<div class="sidemenu"><div class="sidemenu_head">Links</div> <div class="sidemenu_body"><ul><li><a href="#masthead">Top of page</a></li> <li><a href="#run_stats">VEP run statistics</a></li> <li><a href="#gen_stats">General statistics</a></li> <li><a href="#var_class">Variant classes</a></li> <li><a href="#var_cons">Consequences (most severe)</a></li> <li><a href="#consequences">Consequences (all)</a></li> <li><a href="#coding">Coding consequences</a></li> <li><a href="#sift">SIFT summary</a></li> <li><a href="#polyphen">PolyPhen summary</a></li> <li><a href="#chr">Variants by chromosome</a></li> <li><a href="#protein">Position in protein</a></li></ul></div></div><div class='main_content'><h3 id="run_stats">VEP run statistics</h3><table class="stats_table"><tr><td>VEP version (API)</td> <td>87 (87)</td></tr> <tr><td>Cache/Database</td> <td>/peterhouse/tools/bcbio/genomes/Hsapiens/GRCh37/vep/homo_sapiens/87_GRCh37</td></tr> <tr><td>Species</td> <td>homo_sapiens</td></tr> <tr><td>Command line options</td> <td><pre>-i ./VarSelectAnalysisResult_20170524173222/example4_20170524173222_varselect_cleanXY.vcf.gz --dir /peterhouse/tools/bcbio/genomes/Hsapiens/GRCh37/vep --cache --offline --fasta /peterhouse/tools/bcbio/genomes/Hsapiens/GRCh37/seq/GRCh37.fa --port 3337 --fork 4 --everything --total_length --plugin dbNSFP,/peterhouse/lchen/csardas/varselect/db/dbNSFP_3.2a_hg19.txt.gz,Ancestral_allele,AltaiNeandertal,Denisova,Ensembl_geneid,Ensembl_transcriptid,Ensembl_proteinid,SIFT_score,SIFT_converted_rankscore,SIFT_pred,LRT_score,LRT_converted_rankscore,LRT_pred,LRT_Omega,MutationTaster_score,MutationTaster_converted_rankscore,MutationTaster_pred,MutationTaster_model,MutationTaster_AAE,MutationAssessor_UniprotID,MutationAssessor_variant,MutationAssessor_score,MutationAssessor_score_rankscore,MutationAssessor_pred,FATHMM_score,FATHMM_converted_rankscore,FATHMM_pred,PROVEAN_score,PROVEAN_converted_rankscore,PROVEAN_pred,fathmm-MKL_coding_score,fathmm-MKL_coding_rankscore,fathmm-MKL_coding_pred,fathmm-MKL_coding_group,MetaSVM_score,MetaSVM_rankscore,MetaSVM_pred,MetaLR_score,MetaLR_rankscore,MetaLR_pred,Reliability_index,integrated_fitCons_score,integrated_fitCons_score_rankscore,integrated_confidence_value,GM12878_fitCons_score,GM12878_fitCons_score_rankscore,GM12878_confidence_value,H1-hESC_fitCons_score,H1-hESC_fitCons_score_rankscore,H1-hESC_confidence_value,HUVEC_fitCons_score,HUVEC_fitCons_score_rankscore,HUVEC_confidence_value,GERP++_NR,GERP++_RS,GERP++_RS_rankscore,phyloP100way_vertebrate,phyloP100way_vertebrate_rankscore,phyloP20way_mammalian,phyloP20way_mammalian_rankscore,phastCons100way_vertebrate,phastCons100way_vertebrate_rankscore,phastCons20way_mammalian,phastCons20way_mammalian_rankscore,SiPhy_29way_pi,SiPhy_29way_logOdds,SiPhy_29way_logOdds_rankscore,1000Gp3_AC,1000Gp3_AF,1000Gp3_AFR_AC,1000Gp3_AFR_AF,1000Gp3_EUR_AC,1000Gp3_EUR_AF,1000Gp3_AMR_AC,1000Gp3_AMR_AF,1000Gp3_EAS_AC,1000Gp3_EAS_AF,1000Gp3_SAS_AC,1000Gp3_SAS_AF,TWINSUK_AC,TWINSUK_AF,ALSPAC_AC,ALSPAC_AF,ESP6500_AA_AC,ESP6500_AA_AF,ESP6500_EA_AC,ESP6500_EA_AF,ExAC_AC,ExAC_AF,ExAC_Adj_AC,ExAC_Adj_AF,ExAC_AFR_AC,ExAC_AFR_AF,ExAC_AMR_AC,ExAC_AMR_AF,ExAC_EAS_AC,ExAC_EAS_AF,ExAC_FIN_AC,ExAC_FIN_AF,ExAC_NFE_AC,ExAC_NFE_AF,ExAC_SAS_AC,ExAC_SAS_AF,clinvar_rs,clinvar_clnsig,clinvar_trait,Interpro_domain,GTEx_V6_gene,GTEx_V6_tissue,cds_strand,Uniprot_acc_Polyphen2,Uniprot_id_Polyphen2,Uniprot_aapos_Polyphen2,Polyphen2_HDIV_score,Polyphen2_HDIV_rankscore,Polyphen2_HDIV_pred,Polyphen2_HVAR_score,Polyphen2_HVAR_rankscore,Polyphen2_HVAR_pred,Transcript_id_VEST3,Transcript_var_VEST3,VEST3_score,VEST3_rankscore,Eigen-raw,Eigen-phred,Eigen-raw_rankscore,Eigen-PC-raw,Eigen-PC-raw_rankscore,GenoCanyon_score,GenoCanyon_score_rankscore -o ./VarSelectAnalysisResult_20170524173222/VsResult_20170524173222_vep.vcf --vcf --force_overwrite</pre></td></tr> <tr><td>Start time</td> <td>2017-05-24 17:53:39</td></tr> <tr><td>End time</td> <td>2017-05-24 18:23:30</td></tr> <tr><td>Run time</td> <td>1791 seconds</td></tr> <tr><td>Input file (format)</td> <td>./VarSelectAnalysisResult_20170524173222/example4_20170524173222_varselect_cleanXY.vcf.gz (VCF)</td></tr> <tr><td>Output file</td> <td>./VarSelectAnalysisResult_20170524173222/VsResult_20170524173222_vep.vcf <a href="./VarSelectAnalysisResult_20170524173222/VsResult_20170524173222_vep.vcf">[text]</a></td></tr></table><h3 id="gen_stats">General statistics</h3><table class="stats_table"><tr><td>Lines of input read</td> <td>98142</td></tr> <tr><td>Variants processed</td> <td>97918</td></tr> <tr><td>Variants remaining after filtering</td> <td>97918</td></tr> <tr><td>Lines of output written</td> <td>97918</td></tr> <tr><td>Novel / existing variants</td> <td>6275 (6.4%) / 91643 (93.6%)</td></tr> <tr><td>Overlapped genes</td> <td>1262</td></tr> <tr><td>Overlapped transcripts</td> <td>4457</td></tr> <tr><td>Overlapped regulatory features</td> <td>3511</td></tr></table><hr /><h3 id="var_class">Variant classes</h3><div id="var_class_pie" style="width: 800px; height: 200px">&nbsp;</div><div id="var_class_table" style="width: 800px; height: 200px">&nbsp;</div><hr /><h3 id="var_cons">Consequences (most severe)</h3><div id="var_cons_pie" style="width: 800px; height: 400px">&nbsp;</div><div id="var_cons_table" style="width: 800px; height: 200px">&nbsp;</div><hr /><h3 id="consequences">Consequences (all)</h3><div id="consequences_pie" style="width: 800px; height: 400px">&nbsp;</div><div id="consequences_table" style="width: 800px; height: 200px">&nbsp;</div><hr /><h3 id="coding">Coding consequences</h3><div id="coding_pie" style="width: 800px; height: 400px">&nbsp;</div><div id="coding_table" style="width: 800px; height: 200px">&nbsp;</div><hr /><h3 id="sift">SIFT summary</h3><div id="sift_pie" style="width: 800px; height: 200px">&nbsp;</div><div id="sift_table" style="width: 800px; height: 200px">&nbsp;</div><hr /><h3 id="polyphen">PolyPhen summary</h3><div id="polyphen_pie" style="width: 800px; height: 200px">&nbsp;</div><div id="polyphen_table" style="width: 800px; height: 200px">&nbsp;</div><hr /><h3 id="chr">Variants by chromosome</h3><div id="chr_bar" style="width: 800px; height: 200px">&nbsp;</div><div id="chr_table" style="width: 800px; height: 200px">&nbsp;</div><hr /><h3 id="chr_22">Distribution of variants on chromosome 22</h3><div id="chr_22_area" style="width: 800px; height: 200px">&nbsp;</div><hr /><h3 id="protein">Position in protein</h3><div id="protein_bar" style="width: 800px; height: 200px">&nbsp;</div></div>
</div></body>
</html>
